apps/hash-frontend/src/blocks/page/mention-view/mention-display.tsx

```diff
@@ -1,156 +1,328 @@
+  const contentRef = useRef<HTMLDivElement>(null);
 
-        // Once the query loads, either display the found name, or display "Unknown User" if the user doesn't exist in the users array
-        const matchingUser = users.find(
-          (user) => user.entityRecordId.entityId === entityId,
-        );
-
-        if (matchingUser) {
-          return {
-            title: matchingUser.preferredName,
-            href: userHref,
-            icon: "@",
-          };
-        }
+  const [popoverOpen, setPopoverOpen] = useState(false);
 
-        /** @todo - What should the href be here? */
-        return {
-          title: "Unknown User",
-          href: `#`,
-          icon: "@",
-        };
-      }
+  const entity = useMemo(
+    () => (entitySubgraph ? getRoots(entitySubgraph)[0] : undefined),
+    [entitySubgraph],
+  );
 
-      case "page": {
-        const page = pages.find(
-          (potentialPage) =>
-            potentialPage.metadata.recordId.entityId === entityId,
-        );
+  const entityLabel = useMemo(
+    () => (entitySubgraph ? generateEntityLabel(entitySubgraph) : undefined),
+    [entitySubgraph],
+  );
 
-        let pageTitle = "";
+  const getOwnerForEntity = useGetOwnerForEntity();
 
-        // Only set the title to "Page" if the query hasn't returned yet
-        if (pagesLoading && !pages.length) {
-          pageTitle = "Page";
-        } else {
-          // Once the query loads, either display the found title, or display "Unknown Page" if the page doesn't exist in the page array
-          pageTitle = page?.title ?? "Unknown Page";
-        }
+  const entityOwnerShortname = useMemo(() => {
+    if (entity) {
+      const { shortname } = getOwnerForEntity(entity);
 
+      return shortname;
+    }
+  }, [entity, getOwnerForEntity]);
+
+  const entityHref = useMemo(() => {
+    if (entity) {
+      return `/@${entityOwnerShortname}/entities/${extractEntityUuidFromEntityId(
+        entity.metadata.recordId.entityId,
+      )}`;
+    }
+  }, [entity, entityOwnerShortname]);
+
+  const title = useMemo(() => {
+    if (!entitySubgraph) {
+      return undefined;
+    }
+
+    if (
+      mention.kind === "entity" ||
+      mention.kind === "page" ||
+      mention.kind === "user"
+    ) {
+      return entityLabel;
+    } else if (mention.kind === "outgoing-link") {
+      const outgoingLinkAndTargetEntities = getOutgoingLinkAndTargetEntities(
+        entitySubgraph,
+        entityId,
+      ).find(
+        ({ linkEntity: linkEntityRevisions }) =>
+          linkEntityRevisions[0]?.metadata.recordId.entityId ===
+          mention.linkEntityId,
+      );
+
+      const targetEntity = outgoingLinkAndTargetEntities?.rightEntity[0];
+
+      const targetEntityLabel = generateEntityLabel(
+        entitySubgraph,
+        targetEntity,
+      );
+
+      return targetEntityLabel;
+    } else {
+      const propertyTypeBaseUrl = extractBaseUrl(mention.propertyTypeId);
```

Comment:
I would store the property type base URL rather than the id, so that the mention does not break if the entity's type is upgraded to change the version of the property type. Looking for the property type in the subgraph would then have to be by base URL rather than versioned URL.

Reply:
Good idea, I've done this in https://github.com/hashintel/hash/pull/3244/commits/63a3c6b4445f9c728bf43610f8239a85aedb91be

Reply:
Good idea, I've done this in https://github.com/hashintel/hash/pull/3244/commits/63a3c6b4445f9c728bf43610f8239a85aedb91be

Reply:
Good idea, I've done this in https://github.com/hashintel/hash/pull/3244/commits/63a3c6b4445f9c728bf43610f8239a85aedb91be